File: src/main/java/com/example/tdd/money/Money.java
package com.example.tdd.money;

class Money implements Expression {

    protected int amount;

    protected String currency;

    public Money(int amount, String currency) {
        this.amount = amount;
        this.currency = currency;
    }

    public boolean equals(Object object) {
        Money money = (Money) object;
        return amount == money.amount
                && currency().equals(money.currency());
    }

    static Money dollar(int amount) {
        return new Money(amount, "USD");
    }

    static Money franc(int amount) {
        return new Money(amount, "CHF");
    }

    Money times(int mutiplier) {
        return new Money(amount * mutiplier, currency);
    }

    public Money reduce(String to) {
        return this;
    }

    Expression plus(Money addend) {
        return new Sum(this, addend);
    }

    protected String currency() {
        return currency;
    }

    @Override
    public String toString() {
        return amount + " " + currency ;
    }
}

File: src/main/java/com/example/tdd/ktmoney/Money.kt
package com.example.tdd.ktmoney

import java.time.Month

class Money(
    val amount: Int,
    val currency: String
): Expression {

    companion object {
        fun dollar(amount: Int): Money = Money(amount, "USD")

        fun franc(amount: Int): Money = Money(amount, "CHF")
    }

    operator fun plus(addend: Money): Expression {
        return Money(this.amount + addend.amount, currency)
    }

    operator fun times(multiplier: Int): Money {
        return Money(amount * multiplier, currency)
    }

    override fun equals(other: Any?): Boolean {
        if (other !is Money) return false
        return this.amount == other.amount && this.currency == other.currency
    }

    override fun hashCode(): Int {
        return amount
    }

    override fun toString(): String {
        return "Money(amount=$amount, currency='$currency')"
    }
}